File: testing/it/ui-js/src/content/jcr_root/apps/core/wcm/tests/test-cases/Teaser/v1/Teaser.js
/*~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
 ~ Copyright 2018 Adobe
 ~
 ~ Licensed under the Apache License, Version 2.0 (the "License");
 ~ you may not use this file except in compliance with the License.
 ~ You may obtain a copy of the License at
 ~
 ~     http://www.apache.org/licenses/LICENSE-2.0
 ~
 ~ Unless required by applicable law or agreed to in writing, software
 ~ distributed under the License is distributed on an "AS IS" BASIS,
 ~ WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 ~ See the License for the specific language governing permissions and
 ~ limitations under the License.
 ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~*/

;(function(h, $) { // eslint-disable-line no-extra-semi
    "use strict";

    window.CQ.CoreComponentsIT.Teaser.v1 = window.CQ.CoreComponentsIT.Teaser.v1 || {};
    var c                                = window.CQ.CoreComponentsIT.commons;
    var teaser                           = window.CQ.CoreComponentsIT.Teaser.v1;
    var testAssetsPath                   = "/content/dam/core-components";
    var testImagePath                    = testAssetsPath + "/core-comp-test-image.jpg";
    var pretitle                         = "Teaser Pretitle";
    var title                            = "Teaser Title";
    var description                      = "Teaser Description";
    var pageName                         = "teaser-page";
    var pageVar                          = "teaser_page";
    var secondPageName                   = "teaser-second-page";
    var secondPageVar                    = "teaser_second_page";
    var pageDescription                  = "teaser page description";
    var actionText2                      = "Action Text 2";
    var actionExternalLink               = "http://www.adobe.com";
    var actionExternalText               = "Adobe";

    teaser.tcExecuteBeforeTest = function(tcExecuteBeforeTest, teaserRT, pageRT) {
        return new h.TestCase("Create sample content", {
            execBefore: tcExecuteBeforeTest
        })
            .execFct(function(opts, done) {
                c.createPage(c.template, c.rootPage, pageName, pageVar, done, pageRT, pageDescription);
            })
            .execFct(function(opts, done) {
                c.createPage(c.template, c.rootPage, secondPageName, secondPageVar, done, pageRT, pageDescription);
            })

            // create a proxy component
            .execFct(function(opts, done) {
                c.createProxyComponent(teaserRT, c.proxyPath, "proxyPath", done);
            })

            // create a proxy component for an image
            .execFct(function(opts, done) {
                c.createProxyComponent(c.rtImage_v2, c.proxyPath, "imageProxyPath", done);
            })

            .execFct(function(opts, done) {
            // we need to set property for image rendering delegation from  teaser proxy to image proxy
                var data = {};
                data.imageDelegate = h.param("imageProxyPath")(opts);
                c.editNodeProperties(h.param("proxyPath")(opts), data, done);
            })

            .execFct(function(opts, done) {
                c.addComponent(h.param("proxyPath")(opts), h.param(pageVar)(opts) + c.relParentCompPath, "cmpPath", done);
            })
            .navigateTo("/editor.html%" + pageVar + "%.html");
    };

    teaser.tcExecuteAfterTest = function(tcExecuteAfterTest, policyPath, policyAssignmentPath) {
        return new h.TestCase("Clean up after test", {
            execAfter: tcExecuteAfterTest
        })

            // delete the test proxies we created
            .execFct(function(opts, done) {
                c.deleteProxyComponent(h.param("proxyPath")(opts), done);
            })

            .execFct(function(opts, done) {
                c.deleteProxyComponent(h.param("imageProxyPath")(opts), done);
            })

            .execFct(function(opts, done) {
                c.deletePolicy("/teaser", done, policyPath);
            })
            .execFct(function(opts, done) {
                c.deletePolicyAssignment("/teaser", done, policyAssignmentPath);
            })

            .execFct(function(opts, done) {
                c.deletePage(h.param(secondPageVar)(opts), done);
            })

            .execFct(function(opts, done) {
                c.deletePage(h.param(pageVar)(opts), done);
            });
    };

    teaser.testFullyConfiguredTeaser = function(tcExecuteBeforeTest, tcExecuteAfterTest, selectors) {
        return new h.TestCase("Fully configured Teaser", {
            execBefore: tcExecuteBeforeTest,
            execAfter: tcExecuteAfterTest
        })
            .execTestCase(c.tcOpenConfigureDialog("cmpPath"))
            .execFct(function(opts, done) {
                c.openSidePanel(done);
            })
            // filter by test assets path
            .fillInput(selectors.assetFinder.filters.path.textField, testAssetsPath)
            .click(selectors.assetFinder.filters.path.buttonListItem + "[value='" + testAssetsPath + "']")
            // drag'n'drop the test image
            .cui.dragdrop(selectors.editDialog.assetDrag(testImagePath), selectors.editDialog.assetDrop)
            .fillInput(selectors.editDialog.linkURL, "%" + pageVar + "%")
            .fillInput(selectors.editDialog.pretitle, pretitle)
            .fillInput(selectors.editDialog.title, title)
            .fillInput(selectors.editDialog.description, description)
            .execTestCase(c.tcSaveConfigureDialog)
            .assert.isTrue(function() {
                return h.find(selectors.component.image + ' img[src*="' + h.param(pageVar)() +
                    '/_jcr_content/root/responsivegrid/teaser"]', "#ContentFrame").size() === 1;
            })
            .assert.isTrue(function() {
                var selector = selectors.component.pretitle;
                return h.find(selector, "#ContentFrame").html().trim() === pretitle;
            })
            .assert.isTrue(function() {
                var selector = "a" + selectors.component.titleLink + '[href$="' + h.param(pageVar)() + '.html"]';
                return h.find(selector, "#ContentFrame").text() === title;
            })
            .assert.isTrue(function() {
                var selector = selectors.component.description;
                return h.find(selector, "#ContentFrame").html().trim() === description;
            });

    };

    teaser.testInheritedPropertiesTeaser = function(tcExecuteBeforeTest, tcExecuteAfterTest, selectors) {
        return new h.TestCase("Teaser with inherited properties", {
            execBefore: tcExecuteBeforeTest,
            execAfter: tcExecuteAfterTest
        })
            .execTestCase(c.tcOpenConfigureDialog("cmpPath"))
            .click(selectors.editDialog.titleFromPage)
            .click(selectors.editDialog.descriptionFromPage)
            .execFct(function(opts, done) {
                c.openSidePanel(done);
            })
            // filter by test assets path
            .fillInput(selectors.assetFinder.filters.path.textField, testAssetsPath)
            .click(selectors.assetFinder.filters.path.buttonListItem + "[value='" + testAssetsPath + "']")
            // drag'n'drop the test image
            .cui.dragdrop(selectors.editDialog.assetDrag(testImagePath), selectors.editDialog.assetDrop)
            .fillInput(selectors.editDialog.linkURL, "%" + pageVar + "%")
            .execTestCase(c.tcSaveConfigureDialog)
            .assert.isTrue(function() {
                return h.find(selectors.component.image + ' img[src*="' + h.param(pageVar)() +
                    '/_jcr_content/root/responsivegrid/teaser"]', "#ContentFrame").size() === 1;
            })
            .assert.isTrue(function() {
                var selector = "a" + selectors.component.titleLink + '[href$="' + h.param(pageVar)() + '.html"]';
                return h.find(selector, "#ContentFrame").text() === pageName;
            })
            .assert.isTrue(function() {
                var selector = selectors.component.description;
                return h.find(selector, "#ContentFrame").html().trim() === pageDescription;
            });

    };

    teaser.testNoImageTeaser = function(tcExecuteBeforeTest, tcExecuteAfterTest, selectors) {
        return new h.TestCase("Teaser with title, description and without image and link", {
            execBefore: tcExecuteBeforeTest,
            execAfter: tcExecuteAfterTest
        })
            .execTestCase(c.tcOpenConfigureDialog("cmpPath"))
            .execFct(function(opts, done) {
                c.openSidePanel(done);
            })

            .fillInput(selectors.editDialog.title, title)
            .fillInput(selectors.editDialog.description, description)
            .execTestCase(c.tcSaveConfigureDialog)
            .assert.isTrue(function() {
                return h.find(selectors.component.image + ' img[src*="' + h.param(pageVar)() +
                        '/_jcr_content/root/responsivegrid/teaser"]', "#ContentFrame").size() === 0;
            })
            .assert.isTrue(function() {
                var selector = selectors.component.title;
                return h.find(selector, "#ContentFrame").text().trim() === title;
            })
            .assert.isTrue(function() {
                var selector = selectors.component.description;
                return h.find(selector, "#ContentFrame").html().trim() === description;
            });
    };

    teaser.testHideElementsTeaser = function(tcExecuteBeforeTest, tcExecuteAfterTest, selectors, policyName, policyLocation, policyPath, policyAssignmentPath) {
        return new h.TestCase("Hide elements for Teaser", {
            execBefore: tcExecuteBeforeTest,
            execAfter: tcExecuteAfterTest
        })

            .execFct(function(opts, done) {
                var data = {};
                data["jcr:title"] = "New Policy";
                data["sling:resourceType"] = "wcm/core/components/policy/policy";
                data["titleHidden"] = "true";
                data["descriptionHidden"] = "true";

                c.createPolicy(policyName + "/new_policy", data, "policyPath", done, policyPath);
            })

            .execFct(function(opts, done) {
                var data = {};
                data["cq:policy"] = policyLocation + policyName + "/new_policy";
                data["sling:resourceType"] = "wcm/core/components/policies/mapping";

                c.assignPolicy(policyName, data, done, policyAssignmentPath);
            }, { after: 1000 })

            // open the dialog
            .execTestCase(c.tcOpenConfigureDialog("cmpPath"))

            .assert.exist(selectors.editDialog.descriptionFromPage, false)
            .assert.exist(selectors.editDialog.titleFromPage, false);
    };

    teaser.testLinksToElementsTeaser = function(tcExecuteBeforeTest, tcExecuteAfterTest, selectors, policyName, policyLocation, policyPath, policyAssignmentPath) {
        return new h.TestCase("Links to elements for Teaser", {
            execBefore: tcExecuteBeforeTest,
            execAfter: tcExecuteAfterTest
        })

            .execFct(function(opts, done) {
                var data = {};
                data["jcr:title"] = "New Policy";
                data["sling:resourceType"] = "wcm/core/components/policy/policy";
                data["imageLinkHidden"] = "true";
                data["titleLinkHidden"] = "true";

                c.createPolicy(policyName + "/new_policy", data, "policyPath", done, policyPath);
            })

            .execFct(function(opts, done) {
                var data = {};
                data["cq:policy"] = policyLocation + policyName + "/new_policy";
                data["sling:resourceType"] = "wcm/core/components/policies/mapping";

                c.assignPolicy(policyName, data, done, policyAssignmentPath);
            }, { after: 1000 })

            // open the dialog
            .execTestCase(c.tcOpenConfigureDialog("cmpPath"))
            .execFct(function(opts, done) {
                c.openSidePanel(done);
            })
            // filter by test assets path
            .fillInput(selectors.assetFinder.filters.path.textField, testAssetsPath)
            .click(selectors.assetFinder.filters.path.buttonListItem + "[value='" + testAssetsPath + "']")
            // drag'n'drop the test image
            .cui.dragdrop(selectors.editDialog.assetDrag(testImagePath), selectors.editDialog.assetDrop)
            .fillInput(selectors.editDialog.linkURL, "%" + pageVar + "%")
            .execTestCase(c.tcSaveConfigureDialog)

            .assert.isTrue(function() {
                var selector = selectors.component.image + ' img[src*="' + h.param(pageVar)() +
                    '/_jcr_content/root/responsivegrid/teaser"]';
                return h.find(selector, "#ContentFrame").length === 1;
            })
            .assert.isTrue(function() {
                var selector = selectors.component.image + " a";
                return h.find(selector, "#ContentFrame").length === 0;
            })
            .assert.isTrue(function() {
                var selector = "a" + selectors.component.titleLink + '[href$="' + h.param(pageVar)() + '.html"]';
                return h.find(selector, "#ContentFrame").length === 0;
            });
    };

    teaser.testDisableActionsTeaser = function(tcExecuteBeforeTest, tcExecuteAfterTest, selectors, policyName, policyLocation, policyPath, policyAssignmentPath) {
        return new h.TestCase("Disable Actions for Teaser", {
            execBefore: tcExecuteBeforeTest,
            execAfter: tcExecuteAfterTest
        })
            .execFct(function(opts, done) {
                var data = {};
                data["jcr:title"] = "New Policy";
                data["sling:resourceType"] = "wcm/core/components/policy/policy";
                data["actionsDisabled"] = "true";

                c.createPolicy(policyName + "/new_policy", data, "policyPath", done, policyPath);
            })

            .execFct(function(opts, done) {
                var data = {};
                data["cq:policy"] = policyLocation + policyName + "/new_policy";
                data["sling:resourceType"] = "wcm/core/components/policies/mapping";

                c.assignPolicy(policyName, data, done, policyAssignmentPath);
            }, { after: 1000 })

            // open the dialog
            .execTestCase(c.tcOpenConfigureDialog("cmpPath"))
            .execFct(function(opts, done) {
                c.openSidePanel(done);
            })

            .assert.isTrue(function() {
                var actionsEnabled = h.find(selectors.editDialog.actionsEnabled);
                return actionsEnabled.prop("disabled") === true && actionsEnabled.prop("checked") === false;
            });
    };

    teaser.testWithActionsTeaser = function(tcExecuteBeforeTest, tcExecuteAfterTest, selectors) {
        return new h.TestCase("Teaser with Actions", {
            execBefore: tcExecuteBeforeTest,
            execAfter: tcExecuteAfterTest
        })
            // open the dialog
            .execTestCase(c.tcOpenConfigureDialog("cmpPath"))
            .click(selectors.editDialog.titleFromPage)
            .click(selectors.editDialog.descriptionFromPage)
            .execFct(function(opts, done) {
                c.openSidePanel(done);
            })
            // filter by test assets path
            .fillInput(selectors.assetFinder.filters.path.textField, testAssetsPath)
            .click(selectors.assetFinder.filters.path.buttonListItem + "[value='" + testAssetsPath + "']")
            // drag'n'drop the test image
            .cui.dragdrop(selectors.editDialog.assetDrag(testImagePath), selectors.editDialog.assetDrop)
            .click(selectors.editDialog.actionsEnabled)
            .simulate(selectors.editDialog.actionLinkURL + " input[type!='hidden']", "key-sequence",
                { sequence: "%" + pageVar + "%{enter}" })
            .click("button:contains('Add')")
            .simulate(selectors.editDialog.actionLinkURL + ":eq(1) input[type!='hidden']", "key-sequence",
                { sequence: "%" + secondPageVar + "%{enter}" })

            .execTestCase(c.tcSaveConfigureDialog)

            .assert.isTrue(function() {
                return h.find(selectors.component.image + ' img[src*="' + h.param(pageVar)() +
                        '/_jcr_content/root/responsivegrid/teaser"]', "#ContentFrame").size() === 1;
            })

            .assert.isTrue(function() {
                var selector = selectors.component.image + " a";
                return h.find(selector, "#ContentFrame").length === 0;
            })

            .assert.isTrue(function() {
                var selector = "a" + selectors.component.titleLink + '[href$="' + h.param(pageVar)() + '.html"]';
                return h.find(selector, "#ContentFrame").text() === pageName;
            })
            .assert.isTrue(function() {
                var selector = selectors.component.description;
                return h.find(selector, "#ContentFrame").html().trim() === pageDescription;
            })

            .assert.isTrue(function() {
                var selector = selectors.component.actionLink + ":contains('" + pageName + "')";
                return h.find(selector, "#ContentFrame").size() === 1;
            })
            .assert.isTrue(function() {
                var selector = selectors.component.actionLink + ":contains('" + secondPageName + "')";
                return h.find(selector, "#ContentFrame").size() === 1;
            });
    };

    teaser.testWithExternalActionsTeaser = function(tcExecuteBeforeTest, tcExecuteAfterTest, selectors) {
        return new h.TestCase("Teaser with External Actions", {
            execBefore: tcExecuteBeforeTest,
            execAfter: tcExecuteAfterTest
        })
        // open the dialog
            .execTestCase(c.tcOpenConfigureDialog("cmpPath"))
            .execFct(function(opts, done) {
                c.openSidePanel(done);
            })
            // filter by test assets path
            .fillInput(selectors.assetFinder.filters.path.textField, testAssetsPath)
            .click(selectors.assetFinder.filters.path.buttonListItem + "[value='" + testAssetsPath + "']")
            // drag'n'drop the test image
            .cui.dragdrop(selectors.editDialog.assetDrag(testImagePath), selectors.editDialog.assetDrop)
            .click(selectors.editDialog.actionsEnabled)
            .fillInput(selectors.editDialog.actionLinkURL, actionExternalLink)
            .fillInput(selectors.editDialog.actionText, actionExternalText)
            .click("button:contains('Add')")
            .fillInput(selectors.editDialog.actionLinkURL + ":eq(1)", "%" + secondPageVar + "%")
            .fillInput(selectors.editDialog.actionText + ":eq(1)", actionText2)

            .execTestCase(c.tcSaveConfigureDialog)

            .assert.isTrue(function() {
                return h.find(selectors.component.image + ' img[src*="' + h.param(pageVar)() +
                        '/_jcr_content/root/responsivegrid/teaser"]', "#ContentFrame").size() === 1;
            })

            .assert.isTrue(function() {
                var selector = selectors.component.image + " a";
                return h.find(selector, "#ContentFrame").length === 0;
            })

            .assert.isTrue(function() {
                var selector = selectors.component.titleLink;
                return h.find(selector, "#ContentFrame").size() === 0;
            })
            .assert.isTrue(function() {
                var selector = selectors.component.description;
                return h.find(selector, "#ContentFrame").size() === 0;
            })

            .assert.isTrue(function() {
                var selector = selectors.component.actionLink + ":contains('" + actionExternalText + "')";
                return h.find(selector, "#ContentFrame").size() === 1;
            })
            .assert.isTrue(function() {
                var selector = selectors.component.actionLink + ":contains('" + actionText2 + "')";
                return h.find(selector, "#ContentFrame").size() === 1;
            });
    };

    /**
     * Test: Checkbox-Textfield Tuple
     *
     * 1. open the edit dialog
     * 2. switch to the 'Text' tab
     * 3. populate the title tuple textfield
     * 4. open the 'Link & Actions' tab
     * 5. add a link
     * 6. open the 'Text' tab
     * 7. verify the title tuple textfield value has not changed and that the textfield is not disabled
     * 8. set 'Get title from linked page' checkbox, checked
     * 9. verify the title value and disabled state
     * 10. set 'Get title from linked page' checkbox, unchecked
     * 11. verify the title has reverted to its previous user-input value
     */
    teaser.testCheckboxTextfieldTuple = function(tcExecuteBeforeTest, tcExecuteAfterTest, selectors) {
        return new h.TestCase("Checkbox-Textfield Tuple", {
            execBefore: tcExecuteBeforeTest,
            execAfter: tcExecuteAfterTest
        })
            // 1.
            .execTestCase(c.tcOpenConfigureDialog("cmpPath"))

            // 2.
            .click(selectors.editDialog.tabs.text)

            // 3.
            .fillInput(selectors.editDialog.title, title)
            .execFct(function() {
                var $textfield = h.find(selectors.editDialog.title);
                $textfield[0].dispatchEvent(new Event("change"));
            })

            // 4.
            .click(selectors.editDialog.tabs.linkAndActions)

            // 5.
            .fillInput(selectors.editDialog.linkURL, "%" + pageVar + "%")

            // 6.
            .click(selectors.editDialog.tabs.text)

            // 7.
            .asserts.isTrue(function() {
                var $textfield = h.find(selectors.editDialog.title);
                return ($textfield.val() === title) && ($textfield[0].disabled === false);
            })

            // 8.
            .click(selectors.editDialog.titleFromPage)

            // 9.
            .asserts.isTrue(function() {
                var $textfield = h.find(selectors.editDialog.title);
                return ($textfield.val() === "") && ($textfield[0].disabled === true);
            })

            // 10.
            .click(selectors.editDialog.titleFromPage)

            // 11.
            .asserts.isTrue(function() {
                var $textfield = h.find(selectors.editDialog.title);
                return ($textfield.val() === title) && ($textfield[0].disabled === false);
            });
    };
}(hobs, jQuery));

File: bundles/core/src/main/java/com/adobe/cq/wcm/core/components/models/Teaser.java
/*~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
 ~ Copyright 2018 Adobe
 ~
 ~ Licensed under the Apache License, Version 2.0 (the "License");
 ~ you may not use this file except in compliance with the License.
 ~ You may obtain a copy of the License at
 ~
 ~     http://www.apache.org/licenses/LICENSE-2.0
 ~
 ~ Unless required by applicable law or agreed to in writing, software
 ~ distributed under the License is distributed on an "AS IS" BASIS,
 ~ WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 ~ See the License for the specific language governing permissions and
 ~ limitations under the License.
 ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~*/
package com.adobe.cq.wcm.core.components.models;

import java.util.List;

import org.apache.sling.api.resource.Resource;
import org.jetbrains.annotations.Nullable;

import com.adobe.cq.wcm.core.components.commons.link.Link;

/**
 * Defines the {@code Teaser} Sling Model for the {@code /apps/core/wcm/components/teaser} component.
 *
 * @since com.adobe.cq.wcm.core.components.models 12.4.0
 */
public interface Teaser extends Component {

    /**
     * Name of the resource property that defines whether or not the teaser has Call-to-Action elements
     *
     * @since com.adobe.cq.wcm.core.components.models 12.4.0
     */
    String PN_ACTIONS_ENABLED = "actionsEnabled";

    /**
     * Name of the child node where the Call-to-Action elements are stored
     *
     * @since com.adobe.cq.wcm.core.components.models 12.4.0
     */
    String NN_ACTIONS = "actions";

    /**
     * Name of the resource property that stores the Call-to-Action link
     *
     * @since com.adobe.cq.wcm.core.components.models 12.4.0
     */
    String PN_ACTION_LINK = "link";

    /**
     * Name of the resource property that stores the Call-to-Action text
     *
     * @since com.adobe.cq.wcm.core.components.models 12.4.0
     */
    String PN_ACTION_TEXT = "text";

    /**
     * Name of the policy property that defines whether or not Call-to-Actions are disabled
     *
     * @since com.adobe.cq.wcm.core.components.models 12.4.0
     */
    String PN_ACTIONS_DISABLED = "actionsDisabled";

    /**
     * Name of the policy property that defines whether or not the image link is hidden.
     *
     * @since com.adobe.cq.wcm.core.components.models 12.4.0
     */
    String PN_IMAGE_LINK_HIDDEN = "imageLinkHidden";

    /**
     * Name of the policy property that defines whether or not the title is hidden.
     *
     * @since com.adobe.cq.wcm.core.components.models 12.4.0
     */
    String PN_TITLE_HIDDEN = "titleHidden";

    /**
     * Name of the policy property that defines whether or not the pretitle is hidden.
     *
     * @since com.adobe.cq.wcm.core.components.models 12.12.0
     */
    String PN_PRETITLE_HIDDEN = "pretitleHidden";

    /**
     * Name of the policy property that defines whether or not the title link is hidden.
     *
     * @since com.adobe.cq.wcm.core.components.models 12.4.0
     */
    String PN_TITLE_LINK_HIDDEN = "titleLinkHidden";

    /**
     * Name of the resource property that defines whether or not the title value is taken from the linked page.
     *
     * @since com.adobe.cq.wcm.core.components.models 12.4.0
     */
    String PN_TITLE_FROM_PAGE = "titleFromPage";

    /**
     * Name of the policy property that defines whether or not the description is hidden.
     *
     * @since com.adobe.cq.wcm.core.components.models 12.4.0
     */
    String PN_DESCRIPTION_HIDDEN = "descriptionHidden";

    /**
     * Name of the resource property that defines whether or not the description value is taken from the linked page.
     *
     * @since com.adobe.cq.wcm.core.components.models 12.4.0
     */
    String PN_DESCRIPTION_FROM_PAGE = "descriptionFromPage";

    /**
     * Name of the policy property that stores the value for this title's HTML element type.
     *
     * @see #getTitleType()
     * @since com.adobe.cq.wcm.core.components.models 12.4.0
     */
    String PN_TITLE_TYPE = "titleType";

    /**
     * Name of the policy property that defines whether or not the title type is shown.
     *
     * @since com.adobe.cq.wcm.core.components.models 12.16.0
     */
    String PN_SHOW_TITLE_TYPE = "showTitleType";

    /**
     * Checks if the teaser has Call-to-Action elements
     *
     * @return {@code true} if teaser has CTAs, {@code false} otherwise
     * @since com.adobe.cq.wcm.core.components.models 12.4.0
     */
    default boolean isActionsEnabled() {
        return false;
    }

    /**
     * Returns the list of Call-to-Action elements
     *
     * @return the list of CTAs
     * @since com.adobe.cq.wcm.core.components.models 12.4.0
     */
    default List<ListItem> getActions() {
        return null;
    }

    /**
     * Returns the primary link of this teaser.
     *
     * @return the primary link of this teaser
     * @since com.adobe.cq.wcm.core.components.models 12.21.0
     */
    @Nullable
    default Link getLink() {
        return null;
    }

    /**
     * Returns the URL to which this teaser links, if one was defined.
     *
     * @return the URL to which teaser links or {@code null}
     * @since com.adobe.cq.wcm.core.components.models 12.4.0
     * @deprecated Please use {@link #getLink()}
     */
    @Deprecated
    default String getLinkURL() {
        return null;
    }

    /**
     * Returns the image resource for this teaser.
     *
     * @return the image resource for this teaser or {@code null}
     * @since com.adobe.cq.wcm.core.components.models 12.4.0
     */
    default Resource getImageResource() {
        return null;
    }

    /**
     * Checks if the link on the image is hidden.
     *
     * @return {@code true} if link is hidden on the image, {@code false} otherwise
     * @since com.adobe.cq.wcm.core.components.models 12.4.0
     */
    default boolean isImageLinkHidden() {
        return false;
    }

    /**
     * Returns this teaser's pretitle, if one was defined.
     *
     * @return the teaser's pretitle or {@code null}
     * @since com.adobe.cq.wcm.core.components.models 12.12.0
     */
    default String getPretitle() {
        return null;
    }

    /**
     * Returns this teaser's title, if one was defined.
     *
     * @return the teaser's title or {@code null}
     * @since com.adobe.cq.wcm.core.components.models 12.4.0
     */
    default String getTitle() {
        return null;
    }

    /**
     * Checks if the link on the title is hidden.
     *
     * @return {@code true} if link is hidden on the title, {@code false} otherwise
     * @since com.adobe.cq.wcm.core.components.models 12.4.0
     */
    default boolean isTitleLinkHidden() {
        return false;
    }

    /**
     * Returns this teaser's description, if one was defined.
     *
     * @return the teaser's description or {@code null}
     * @since com.adobe.cq.wcm.core.components.models 12.4.0
     */
    default String getDescription() {
        return null;
    }

    /**
     * Returns the HTML element type (h1-h6) used for the title.
     *
     * @return the element type
     * @since com.adobe.cq.wcm.core.components.models 12.4.0
     */
    default String getTitleType() {
        return null;
    }

}
